### PR TITLE
Use `payment_method_domains` endpoint for domain registration.

### LINKED ISSUE #9098: [ECE] Apple Pay session cannot be created for Express Checkout Element unless domain is registered with new endpoint
### Describe the bug

When clicking the Apple Pay button with the `_wcpay_feature_stripe_ece` option set to `1`, the Apple Pay dialog closes and an error is thrown in the console:

```
Could not create Apple Pay session. Please make sure you have registered this Stripe account. For more information, see https://stripe.com/docs/apple-pay#web.
``` 

This is because we need to register the domain first with this new endpoint: https://api.stripe.com/v1/payment_method_domains ([src](https://docs.stripe.com/api/payment_method_domains/create)).

We might want to consider setting `applePay` to `auto` instead of `always` when rendering the button to make sure it's not available in case the domain verification has failed or the button shouldn't be displayed for some reason.

To fix this, we can update or add a new endpoint to the server and trigger domain registrations when WooPayments gets updated:
https://github.com/Automattic/woocommerce-payments/blob/ff072f5f1e28be9c7667d1c484d62f93b629d2c7/includes/class-wc-payments-apple-pay-registration.php#L83

### To Reproduce

1. Enable the ECE flag.
2. Change your domain with JT/ngrok.
3. As a shopper on Safari, try to checkout with Apple Pay.
4. Notice the Apple Pay modal opens up, but closes shortly after.

### PR DESCRIPTION
**Pull request by rafaelzaleski**
Fixes #9098 

#### Changes proposed in this Pull Request

Use the `payment_method_domains` endpoint for domain registration, a requirement for processing Apple Pay payments in Express Checkout Elements (ECE).

#### Testing instructions

**Pre-requisites:**
- Checkout to branch `add/domain-registration-controller` on the server.
- Point WooPayments to your local server on the above branch using the dev tools.
- Expose your test site to the internet and ensure it's served over HTTPS. I suggest using ngrok since it'll generate a random domain and you're testing the domain registration.
- Do not checkout to this branch yet.

**On the `develop` branch:**
- `npm run build:client`
- Ensure Apple Pay is enabled in the WooPayments settings. Ensure it's enabled on all pages (product, cart, checkout).
- Go to the product page. The Apple Pay button should not be visible.

**On the `add/9098-domain-registration-endpoint-update` branch:**

**Test that it's possible to process Apple Pay payments in the new domain:**
- `npm run build:client`
- Access the WooPayments settings. Disable Express Checkout for Apple Pay and Google Pay. Save. Enable them again. This should register the domain.
- Repeat the test from the `develop` branch but ensure the Apple Pay button is visible.

**Check the logs for the `payment_method_domains` endpoint request:**
- Go to the WC Logs `/wp-admin/admin.php?page=wc-status&tab=logs`.
- Click on the latest WooPayments log.
- Search for `payment_method_domains`.
- Review the server response and ensure it has the correct format:

```plaintext
array (
  'id' => 'pmd_xxxxxxxxxxxxx',
  'object' => 'payment_method_domain',
  'apple_pay' => 
  array (
    'status' => 'active',
  ),
  'created' => {timestamp},
  'domain_name' => {website_domain},
  'enabled' => true,
  'google_pay' => 
  array (
    'status' => 'active',
  ),
  'link' => 
  array (
    'status' => 'active',
  ),
  'livemode' => false,
  'paypal' => 
  array (
    'status' => 'active',
  ),
)
```

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9139` or branch name `add/9098-domain-registration-endpoin-update` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/9098-domain-registration-endpoin-update) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 5845ee2286f1f5f6c5b7ac145b116fe77f945397
- Build time: 2024-07-24 17:14:48 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payments-apple-pay-registration.php [hunk 1]
```diff
@@ -313,9 +313,9 @@ public function register_domain_with_apple() {
 	 * Process the Apple Pay domain verification if proper settings are configured.
 	 */
 	public function verify_domain_if_configured() {
-		// If Payment Request Buttons are not enabled, or account is not live,
+		// If Payment Request Buttons are not enabled,
```
Comment (asumaran) [inline includes/class-wc-payments-apple-pay-registration.php]: Is this method used for PRB and ECE? If not maybe we change PRB to "Express Checkout Elements"?

Comment (rafaelzaleski) [reply includes/class-wc-payments-apple-pay-registration.php]: Updated the comment on https://github.com/Automattic/woocommerce-payments/pull/9139/commits/48bf9073c605d155a9074cc67fb5f1cb5538ab65

Comment (asumaran) [reply includes/class-wc-payments-apple-pay-registration.php]: awesome that reads even better.

File: includes/wc-payment-api/class-wc-payments-api-client.php [hunk 2]
```diff
@@ -1603,22 +1603,25 @@ public function link_session_to_customer( $session_id, $customer_id ) {
...
 	 *
 	 * @throws API_Exception If an error occurs.
 	 */
-	public function register_domain_with_apple( $domain_name ) {
+	public function register_domain( $domain_name ) {
 		return $this->request(
 			[
-				'test_mode'   => false, // Force live mode - Domain registration doesn't work in test mode.
 				'domain_name' => $domain_name,
 			],
```
Comment (reykjalin) [inline includes/wc-payment-api/class-wc-payments-api-client.php]: I think we should also explicitly [enable the domain](https://docs.stripe.com/api/payment_method_domains/create#create_payment_method_domain-enabled) by sending `'enabled' => true,` here. Enabling the GPay/APay buttons in WooPayments is explicitly trying to enable the buttons, and thus not only is the domain registration required, but also actually enabling the domain.

```php
			[
				'domain_name' => $domain_name,
				// ⛔️ For some reason this needs to be a string. If it's an actual boolean it's serialized as a 1 (integer) when forwarded by the server causing an invalid request error being returned from Stripe.
				'enabled' => 'true',
			],
```

With a corresponding change to allow the `enabled` parameter on the server.

You can test that this is working as intended by:

```sh
# Find the payment method domain ID you want to disable from the list of available domains.
# Looks like `pmd_abc123`.
curl https://api.stripe.com/v1/payment_method_domains \
  -u "<secret_key>:" \
  -H "Stripe-Account: <acct_id>"

# Disable the domain.
curl https://api.stripe.com/v1/payment_method_domains/<pmd_abc123> \
  -u "<secret_key>:" \
  -H "Stripe-Account: <acct_id>" \
  -d enabled=false

```

Then go disable Apple Pay/Google Pay, and re-enable them.

File: client/express-checkout/blocks/components/express-checkout-preview.js [hunk 3]
```diff
@@ -62,8 +62,8 @@ export const ExpressCheckoutPreviewComponent = ( {
 		},
 		paymentMethods: {
 			link: 'never',
-			googlePay: 'always',
-			applePay: 'always',
+			googlePay: 'auto',
+			applePay: 'auto',
```
Comment (reykjalin) [inline client/express-checkout/blocks/components/express-checkout-preview.js:66]: Thanks for making this change ❤️

**Review by reykjalin — CHANGES_REQUESTED** [review]
Looks good and tests well! I had one suggestion for including an extra parameter in the request, but otherwise this look fantastic! 🎉

File: includes/wc-payment-api/class-wc-payments-api-client.php (hunk 2 above)
Comment (rafaelzaleski) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: Thank you for the suggestion. Added on https://github.com/Automattic/woocommerce-payments/pull/9139/commits/8a5f32ef9f2d96a5fa1ba337479eb2ed019166ae

**Review by reykjalin — APPROVED** [review]
Looking good! 🚀

**Pull request merged**